src/Api/Upload.php

```diff
@@ -0,0 +1,170 @@
+			static::validateChunk($tmpWithId, $filename, $offset, $id);
+
+			// stream chunk content and append it to partial file
+			stream_copy_to_stream(
+				fopen($source, 'r'),
+				fopen($tmpWithId, 'a')
+			);
+
+			// clear file stat cache so the following call to `F::size`
+			// really returns the updated file size
+			clearstatcache();
+
+			// if partial file isn't complete yet, return early
+			if (F::size($tmpWithId) < (int)$total) {
+				return null;
+			}
+
+			// remove id from partial filename now the file is complete,
+			// so we can pass the path from the tmp upload directory
+			// as new source path for the file back to the API upload method
+			rename($tmpWithId, $tmp);
+
+			return $tmp;
+		}
+
+		return $source;
+	}
+
+	/**
+	 * Returns the ideal size for a file chunk
+	 */
+	public static function chunkSize(): int
+	{
+		$max = [
+			Str::toBytes(ini_get('upload_max_filesize')),
+			Str::toBytes(ini_get('post_max_size'))
+		];
+
+		// if server is behind a cloudflare proxy
+		if (isset($_SERVER['HTTP_CF_CONNECTING_IP']) === true) {
+			$max[] = Str::toBytes('100M');
+		}
+
+		// to be sure, only use 95% of the max possible upload size
+		return (int)floor(min($max) * 0.95);
+	}
+
+	/**
+	 * Clean up tmp directory of stale files
+	 */
+	public static function clean(): void
+	{
+		foreach (Dir::files($dir = static::tmp(), [], true) as $file) {
+			// remove any file that hasn't been altered
+			// in the last 24 hours
+			if (F::modified($file) < time() - 86400) {
+				F::remove($file);
+			}
+		}
+
+		if (Dir::isEmpty($dir) === true) {
+			Dir::remove($dir);
+		}
+	}
+
+	/**
+	 * Returns root of directory used for
+	 * temporarily storing (incomplete) uploads
+	 * @codeCoverageIgnore
+	 */
+	protected static function tmp(): string
+	{
+		return App::instance()->root('cache') . '/.uploads';
+	}
+
+	/**
+	 * Ensures the sent chunk is valid
+	 *
+	 * @throws \Kirby\Exception\DuplicateException Duplicate first chunk (same filename and id)
+	 * @throws \Kirby\Exception\Exception Chunk offset does not match existing tmp file
+	 * @throws \Kirby\Exception\NotFoundException Subsequent chunk has no  existing tmp file
+	 */
+	protected static function validateChunk(
+		string $tmp,
+		string $filename,
+		int $offset,
+		string $id
+	): void {
+		if (F::exists($tmp) === true) {
+			if ($offset === 0) {
+				throw new DuplicateException('A tmp file upload with the same filename "' . $filename . '" and upload id "' . $id . '" already exists');
+			}
+
+			if ($offset !== F::size($tmp)) {
+				throw new Exception('Chunk offset ' . $offset . ' does not match the existing tmp upload file size of ' . F::size($tmp));
+			}
+		} elseif ($offset > 0) {
+			throw new NotFoundException('Chunk offset ' . $offset . ' for non-existing tmp file "' . $filename . '"');
+		}
+
+		// validate file
+		$parent = new Page(['slug' => 'tmp']);
+		$file   = new File(['filename' => $filename, 'parent' => $parent]);
+		FileRules::validFile($file, false);
```

Comment:
@lukasbestle Does it make sense to run these on every chunk request or should we limit it to only the first?

Reply:
I think that would be fine. As we only validate the filename and extension, we can assume that this check already happened for files that exist at the `$tmp` path.